### ISSUE TITLE
Likely incorrect code in spend validation

### ISSUE THREAD

**Issue by daira** [zcoinofficial/zcoin#71]
https://github.com/zcoinofficial/zcoin/blob/master/src/main.cpp#L835
```
if (pubcoinId < 1 && pubcoinId == INT_MAX) { // IT BEGINS WITH 1
    return state.DoS(100, error("CTransaction::CheckTransaction() : Error: nSequence is not correct format"));
}
```
The condition cannot be true.

Some of the other code around this point probably needs more thorough review; there are several stale comments, for example.

**Comment by nopslide** [comment]
Thank you, it should be ||

**Comment by daira** [comment]
Note that the whole of the validation code for Zerocoin spends is duplicated for each denomination. The copies for denominations other than ``ZQ_LOVELACE`` have just ``pubCoin < 1`` in the corresponding condition. Why would anyone wire it this way -- why not loop over the denominations?

**Comment by daira** [comment]
The stake comments seem to have come from Tutorial.cpp in libzerocash.

**Comment by nopslide** [comment]
@daira it should come from libzerocoin ?

**Comment by bitcartel** [comment]
I left a comment on a PR that has already been merged, so copying here.

There's a risk of an integer overflow with pubcoinId as nSequence is an unsigned int.

```
 int pubcoinId = txin.nSequence;
...
 if (pubcoinId < 1 && pubcoinId == INT_MAX) { // IT BEGINS WITH 1
```

```
class CTxIn
...
    unsigned int nSequence;
```

**Comment by daira** [comment]
The main issue I pointed out was fixed in 0359bcb2ead7feac5a3a5f6c6581b1b54cda8dd1, but the integer overflow @bitcartel mentioned has not been fixed, the validation code is still duplicated per denomination, and [the stale comments are still there](https://github.com/zcoinofficial/zcoin/blob/a6148876ae016f2627b8047611cdfe1e23a2a5af/src/main.cpp#L821-L822).

**Comment by riordant** [comment]
Closing as all issues raised in this thread have been rectified.
